sirepo/package_data/static/js/sirepo.js

```diff
@@ -1556,14 +1544,157 @@ SIREPO.app.service('validationService', function(utilities) {
 
 });
 
+SIREPO.app.factory('srCache', function(appState, $rootScope) {
+
+    // Browser side caching implemented using indexedDB
+    // - Caches sim frame responses
+    // - Allows cache to be cleared by simId or (simId, modelName)
+    // - Keeps updateTime on item access and deletes expired records at startup
+
+    const self = {};
+    const STORE = 'db';
+    const FRAME = 'frame';
+    // 30 days until expired
+    const EXPIRY_TIME = 30 * 24 * 60 * 60 * 1000;
+    let db = null;
+
+    const deleteKeys = (keys) => {
+        if (! keys.length) {
+            return;
+        }
+        withObjectStore('readwrite', (o) => {
+            for (const k of keys) {
+                o.delete(k);
+            }
+        });
+    };
+
+    const initializeDatabase = () => {
+        if (! window.indexedDB || ! SIREPO.authState.uiWebSocket) {
+            return;
+        }
+        const r = window.indexedDB.open('srCache', 2);
+        r.onsuccess = (event) => {
+            db = event.target.result;
+            removeOldRecords();
+        };
+        r.onupgradeneeded = (event) => {
+            db = event.target.result;
+            if (db.objectStoreNames.contains(STORE)) {
+                db.deleteObjectStore(STORE);
+            }
+            const o = db.createObjectStore(STORE);
+            o.createIndex('simId', '_srcache_simId', { unique: false });
+            o.createIndex('updateTime', '_srcache_updateTime', { unique: false });
+        };
+    };
+
+    const invokeCallback = (callback, value) => {
+        $rootScope.$applyAsync(() => callback(value));
+    };
+
+    const objectKey = (prefix, value) => prefix + ':' + value;
+
+    const objectStore = (mode) => {
```

Comment:
In past projects we've used an "unsafe" prefix for functions which might return null, so it could be ```unsafeObjectStore```. I'm not fond of the "maybe" prefix. In pykern, "unchecked" isn't documenting the return value, but that the function won't throw an exception on failure.

I couldn't find any good references for how to mark a method name which might return null with javascript. Javascript seems to be "unsafe" by default - most of the library functions like getElementById() or getSelection() can return null.

The function could be renamed ```getObjectStore``` and documentation added to the function stating that it will return null if the object store is not available.

Reply:
> In pykern, "unchecked" isn't documenting the return value, but that the function won't throw an exception on failure.

In sirepo it documents the return value (ex [unchecked_get_user](https://github.com/radiasoft/sirepo/blob/7a5b5d08b04fb6175d76fd454bfdacb7c34afdd4/sirepo/auth/__init__.py#L581). 

> getObjectStore

get or unsafe (or unchecked) are all good with me.

Reply:
Yes, it looks like there are examples in sirepo of unchecked for null returns. Also ```unchecked_search_by```. 

For now I've changed it to ```getObjectStore``` and added the doc.